### PR TITLE
Implement neighbor sampler

### PR DESCRIPTION
**Pull request by kgajdamo**
The purpose of this PR is to implement neighbor sampler in pyg-lib.

What has been done:
- [x] Registered neighbor_sampler as a torch custom operator.
- [x] Moved neighbor_sampler implementation from pytorch_sparse to pyg-lib.
- [x] pyg-lib neighbor_sampler supports CSR matrix format.
- [x] Added simple test (test passed).

In progress:
- [x] Replace to_local_node map with Mapper.
- [x] Add support for scalar types.
- [ ] Add support for isolated and return_edge_id arguments.

### CODE REVIEW

File: pyg_lib/sampler/__init__.py [hunk 1]
```diff
@@ -4,6 +4,41 @@
 from torch import Tensor
 
 
+def neighbor_sample(
+    rowptr: Tensor, col: Tensor, seed: Tensor, num_neighbors: list[int],
+    replace: bool = False, directed: bool = True, isolated: bool = True,
+    return_edge_id: bool = True
```
Comment (rusty1s) [inline pyg_lib/sampler/__init__.py:10]:
```suggestion
    return_edge_id: bool = True,
```
and let `yapf` format.

File: pyg_lib/sampler/__init__.py [hunk 2]
```diff
@@ -4,6 +4,41 @@
...
+) -> Tuple[Tensor, Tensor, Tensor, Optional[Tensor]]:
+    r"""Recursively samples neighbors from all nodes indices in :obj:`seed`
```
Comment (rusty1s) [inline pyg_lib/sampler/__init__.py:12]:
```suggestion
    r"""Recursively samples neighbors from all node indices in :obj:`seed`
```

File: pyg_lib/sampler/__init__.py [hunk 3]
```diff
@@ -4,6 +4,41 @@
...
+    in the graph given by :obj:`(rowptr, col)`.
+
+    Args:
+        rowptr (torch.Tensor): Compressed source node indices.
+        col (torch.Tensor): Target node indices.
+        seed (torch.Tensor):
```
Comment (rusty1s) [inline pyg_lib/sampler/__init__.py:18]:
```suggestion
        seed (torch.Tensor): The seed node indices.
```

File: pyg_lib/sampler/__init__.py [hunk 4]
```diff
@@ -4,6 +4,41 @@
...
+        num_neighbors (list[int]): The number of neighbors to sample for each
+            node in each iteration. In heterogeneous graphs, may also take in a
+            dictionary denoting the amount of neighbors to sample for each
+            individual edge type. If an entry is set to :obj:`-1`, all neighbors
+            will be included.
+        replace (bool, optional): If set to :obj:`True`, will sample with
+            replacement. (default: :obj:`False`)
+        directed (bool, optional): If set to :obj:`False`, will include all
+            edges between all sampled nodes. (default: :obj:`True`)
+        isolated (bool, optional): #TODO(kgajdamo): add description
```
Comment (rusty1s) [inline pyg_lib/sampler/__init__.py:28]: Let's rename to `disjoint`: "If set to :obj:`True`, will create disjoint subgraphs for every seed node. (default: :obj:`False`)

File: test/csrc/sampler/test_neighbor_sample.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import torch
```
Comment (rusty1s) [inline test/csrc/sampler/test_neighbor_sample.py:1]: It would be great if we can just test in C++. WDYT?

File: pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp [hunk 6]
```diff
@@ -1,21 +1,147 @@
...
                        const at::Tensor& col,
                        const at::Tensor& seed,
-                       const std::vector<int64_t> num_neighbors,
+                       const std::vector<int64_t>& num_neighbors,
                        bool replace,
                        bool directed,
                        bool isolated,
                        bool return_edge_id) {
-  return std::make_tuple(rowptr, col, seed, at::nullopt);
+  if (replace && directed) {
```
Comment (rusty1s) [inline pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp:136]: Raise an error in case `isolated` is set.

Comment (rusty1s) [inline pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp:136]: Raise an error in case `return_edge_id=False`?

File: pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp [hunk 7]
```diff
@@ -1,21 +1,147 @@
...
 #include <torch/library.h>
+#include <torch/script.h>
+
+#include "../neighbor.h"
+#include "pyg_lib/csrc/sampler/cpu/mapper.h"
+#include "utils.h"
+
+#include "parallel_hashmap/phmap.h"
+
+#ifdef _WIN32
```
Comment (rusty1s) [inline pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp:11]: I don't think we need this anymore.

File: pyg_lib/csrc/sampler/cpu/utils.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <torch/torch.h>
+
+#include "parallel_hashmap/phmap.h"
+
+template <typename scalar_t>
+inline torch::Tensor from_vector(const std::vector<scalar_t>& vec,
```
Comment (rusty1s) [inline pyg_lib/csrc/sampler/cpu/utils.h:8]: This already exists in `cpu/convert.h` - might be good to re-use.

File: pyg_lib/csrc/sampler/cpu/utils.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+inline c10::Dict<key_t, torch::Tensor> from_vector(
+    const phmap::flat_hash_map<key_t, std::vector<scalar_t>>& vec_dict,
+    bool inplace = false) {
+  c10::Dict<key_t, torch::Tensor> out_dict;
+  for (const auto& kv : vec_dict)
+    out_dict.insert(kv.first, from_vector<scalar_t>(kv.second, inplace));
+  return out_dict;
+}
+
+inline int64_t uniform_randint(int64_t low, int64_t high) {
```
Comment (rusty1s) [inline pyg_lib/csrc/sampler/cpu/utils.h:26]: At best, we can use `cpu/random/rand_engine` for this and drop this functions altogether. See [here](https://github.com/pyg-team/pyg-lib/blob/master/test/csrc/random/test_rand_engine.cpp#L9) for an example.

**Review by rusty1s — COMMENTED** [review]
Thanks for getting us started! This looks pretty good. Would prefer moving basic tests to C++, and using `rand_engine` for any random number generator logic.

File: test/csrc/sampler/test_neighbor_sample.py (hunk 5 above)
Comment (kgajdamo) [reply test/csrc/sampler/test_neighbor_sample.py:1]: Sure, but unfortunately, when I want to compile pyg-lib with unit tests I am receiving undefined reference error. I wonder if this is only happening on my side?

Comment (rusty1s) [reply test/csrc/sampler/test_neighbor_sample.py:1]: Can you show me the log?

Comment (kgajdamo) [reply test/csrc/sampler/test_neighbor_sample.py:1]: /home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/test_biased_random.dir/test/csrc/random/test_biased_random.cpp.o: in function `BiasedSamplingCDFConversionTest_BasicAssertions_Test::TestBody()':
test_biased_random.cpp:(.text._ZN52BiasedSamplingCDFConversionTest_BasicAssertions_Test8TestBodyEv+0x344): undefined reference to `testing::internal::GetBoolAssertionFailureMessage[abi:cxx11](testing::AssertionResult const&, char const*, char const*, char const*)'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: test_biased_random.cpp:(.text._ZN52BiasedSamplingCDFConversionTest_BasicAssertions_Test8TestBodyEv+0x479): undefined reference to `testing::internal::GetBoolAssertionFailureMessage[abi:cxx11](testing::AssertionResult const&, char const*, char const*, char const*)'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/test_biased_random.dir/test/csrc/random/test_biased_random.cpp.o: in function `BiasedSamplingAliasConversionTest_BasicAssertions_Test::TestBody()':
test_biased_random.cpp:(.text._ZN54BiasedSamplingAliasConversionTest_BasicAssertions_Test8TestBodyEv+0x90b): undefined reference to `testing::internal::GetBoolAssertionFailureMessage[abi:cxx11](testing::AssertionResult const&, char const*, char const*, char const*)'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: test_biased_random.cpp:(.text._ZN54BiasedSamplingAliasConversionTest_BasicAssertions_Test8TestBodyEv+0x9e3): undefined reference to `testing::internal::GetBoolAssertionFailureMessage[abi:cxx11](testing::AssertionResult const&, char const*, char const*, char const*)'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: test_biased_random.cpp:(.text._ZN54BiasedSamplingAliasConversionTest_BasicAssertions_Test8TestBodyEv+0xabd): undefined reference to `testing::internal::GetBoolAssertionFailureMessage[abi:cxx11](testing::AssertionResult const&, char const*, char const*, char const*)'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/test_biased_random.dir/test/csrc/random/test_biased_random.cpp.o:test_biased_random.cpp:(.text._ZN54BiasedSamplingAliasConversionTest_BasicAssertions_Test8TestBodyEv+0xb9d): more undefined references to `testing::internal::GetBoolAssertionFailureMessage[abi:cxx11](testing::AssertionResult const&, char const*, char const*, char const*)' follow
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/test_biased_random.dir/test/csrc/random/test_biased_random.cpp.o: in function `testing::AssertionResult testing::internal::CmpHelperOpFailure<int, double>(char const*, char const*, int const&, double const&, char const*)':
test_biased_random.cpp:(.text._ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_[_ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_]+0xa4): undefined reference to `testing::Message::GetString[abi:cxx11]() const'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: test_biased_random.cpp:(.text._ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_[_ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_]+0x15d): undefined reference to `testing::Message::GetString[abi:cxx11]() const'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: test_biased_random.cpp:(.text._ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_[_ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_]+0x1ff): undefined reference to `testing::Message::GetString[abi:cxx11]() const'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: test_biased_random.cpp:(.text._ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_[_ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_]+0x2b0): undefined reference to `testing::Message::GetString[abi:cxx11]() const'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: test_biased_random.cpp:(.text._ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_[_ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_]+0x352): undefined reference to `testing::Message::GetString[abi:cxx11]() const'
/home/kgajdamo/miniconda3/envs/pyg/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: CMakeFiles/test_biased_random.dir/test/csrc/random/test_biased_random.cpp.o:test_biased_random.cpp:(.text._ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_[_ZN7testing8internal18CmpHelperOpFailureIidEENS_15AssertionResultEPKcS4_RKT_RKT0_S4_]+0x40b): more undefined references to `testing::Message::GetString[abi:cxx11]() const' follow
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/test_biased_random.dir/build.make:102: test_biased_random] Error 1
make[1]: *** [CMakeFiles/Makefile2:204: CMakeFiles/test_biased_random.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

Comment (rusty1s) [reply test/csrc/sampler/test_neighbor_sample.py:1]: @ZenoTan Did you have seen this before? If not, let's ignore C++ tests for now :)

Comment (ZenoTan) [reply test/csrc/sampler/test_neighbor_sample.py:1]: Didn't meet this but it looks like the linkage to gtest failed. It's fine to ignore though.

Comment (kgajdamo) [reply test/csrc/sampler/test_neighbor_sample.py:1]: Do you also get this error during compilation or is it just me?

File: pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp (hunk 6 above)
Comment (kgajdamo) [reply pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp:136]: At the moment it always should raise an error?

Comment (rusty1s) [reply pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp:136]: Can you clarify? The argument is ignored currently, while `sample` implements the `isolated=False` case.

Comment (kgajdamo) [reply pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp:144]: Ok, I have added an error. Please let me know if this is what You have on mind.

File: pyg_lib/sampler/__init__.py (hunk 1 above)
Comment (kgajdamo) [reply pyg_lib/sampler/__init__.py]: Done.

File: pyg_lib/sampler/__init__.py (hunk 3 above)
Comment (kgajdamo) [reply pyg_lib/sampler/__init__.py]: Thanks, description added.

File: pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp (hunk 7 above)
Comment (kgajdamo) [reply pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp]: Removed.

File: pyg_lib/csrc/sampler/cpu/utils.h (hunk 9 above)
Comment (kgajdamo) [reply pyg_lib/csrc/sampler/cpu/utils.h]: Done.

File: pyg_lib/sampler/__init__.py (hunk 2 above)
Comment (kgajdamo) [reply pyg_lib/sampler/__init__.py]: Done.

File: pyg_lib/sampler/__init__.py (hunk 4 above)
Comment (kgajdamo) [reply pyg_lib/sampler/__init__.py]: Done.

File: test/csrc/sampler/test_neighbor_sample.py (hunk 5 above)
Comment (ZenoTan) [reply test/csrc/sampler/test_neighbor_sample.py:1]: I did not get exactly the same, but similar issues.

Comment (dszwicht) [reply test/csrc/sampler/test_neighbor_sample.py:1]: @kgajdamo @rusty1s  I noticed this error occurs when we have a mix of CF/anaconda c++ packages in our toolchain. Compiling this code in simple [venv](https://docs.python.org/3/library/venv.html) works well.

**Review by rusty1s — APPROVED** [review]

**Comment by rusty1s** [comment]
I think this is a great start. Thank you so much for putting this together. I also added `scalar_t` support.
